File: src/opusmanager/layer_flag.py
""" Class definition of requisite component classes of the Flag Layer. """
from __future__ import annotations
from typing import Optional, Dict, List, Tuple, Any

from .miditree import MIDITree
from .layer_base import OpusManagerBase, BeatKey

class UpdatesCache:
    """
        Dict of changes with a few ease-of-use functions.
        Used by FlagLayer to cache the changes to the OpusManager
        such so a GUI can more quickly and easily update itself.
    """
    def __init__(self):
        self.cache: Dict = {}

    def flag(self, key: str, value: Any) -> None:
        """Flag a change, caching the specifics."""
        if key not in self.cache:
            self.cache[key] = []
        self.cache[key].append(value)

    def unflag(self, key: str, value: Any) -> None:
        """Cancel a flagged changed."""
        if key not in self.cache:
            return
        self.cache[key].remove(value)

    def fetch(self, key: str, noclobber: bool = False) -> List[Any]:
        """
            Get the flagged changes as a list.
            Clears out the list unless *noclobber* is True
        """
        output = self.cache.get(key, [])
        if not noclobber:
            self.cache[key] = []
        return output

class FlagLayer(OpusManagerBase):
    """
        Flags changes made to the OpusManager
        so a GUI can update itself accordingly
    """
    def __init__(self):
        super().__init__()
        self.updates_cache = UpdatesCache()

    ## Layer-Specific Functions
    def flag(self, key: str, value: Any) -> None:
        """Wrapper for UpdatesCache's flag() method"""
        self.updates_cache.flag(key, value)

    def fetch(self, key: str, noclobber: bool = False) -> List[Any]:
        """Wrapper for UpdatesCache's fetch() method"""
        return self.updates_cache.fetch(key, noclobber)

    ## OpusManagerBase Functions
    def replace_tree(
            self,
            beat_key: BeatKey,
            position: List[int],
            tree: MIDITree) -> None:
        super().replace_tree(beat_key, position, tree)
        self.flag('beat_change', beat_key)

    def overwrite_beat(
            self,
            old_beat: BeatKey,
            new_beat: BeatKey) -> None:
        super().overwrite_beat(old_beat, new_beat)
        self.flag('beat_change', old_beat)

    def unlink_beat(self, beat_key: BeatKey) -> None:
        super().unlink_beat(beat_key)
        self.flag('beat_change', beat_key)

    def link_beats(
            self,
            beat: BeatKey,
            target: BeatKey) -> None:
        super().link_beats(beat, target)
        self.flag('beat_change', beat)

    def insert_after(self, beat_key: BeatKey, position: List[int]) -> None:
        super().insert_after(beat_key, position)
        self.flag('beat_change', beat_key)

    def split_tree(
            self,
            beat_key: BeatKey,
            position: List[int],
            splits: int) -> None:
        super().split_tree(beat_key, position, splits)
        self.flag('beat_change', beat_key)

    def swap_channels(self, channel_a: int, channel_b: int) -> None:
        super().swap_channels(channel_a, channel_b)

        len_a = len(self.channel_trees[channel_a])
        len_b = len(self.channel_trees[channel_b])
        for i in range(len_b):
            self.flag('line', (channel_a, len_b - 1 - i, 'pop'))

        for i in range(len_a):
            self.flag('line', (channel_b, len_a - 1 - i, 'pop'))

        for i in range(len_a):
            self.flag('line', (channel_a, i, 'new'))

        for i in range(len_b):
            self.flag('line', (channel_b, i, 'new'))

    def _new(self) -> None:
        super()._new()

        for i in range(self.opus_beat_count):
            self.flag('beat', (i, 'new'))

        for i, channel in enumerate(self.channel_trees):
            for j, _line in enumerate(channel):
                self.flag('line', (i, j, 'init'))

    def _load(self, path: str) -> None:
        super()._load(path)

        for i in range(self.opus_beat_count):
            self.flag('beat', (i, 'new'))

        for i, channel in enumerate(self.channel_trees):
            for j, _line in enumerate(channel):
                self.flag('line', (i, j, 'init'))

    def set_event(
            self,
            beat_key: BeatKey,
            position: List[int],
            value: int,
            *,
            relative: bool = False) -> None:
        super().set_event(beat_key, position, value, relative=relative)
        self.flag('beat_change', beat_key)

    def set_percussion_event(
            self,
            beat_key: BeatKey,
            position: List[int]) -> None:
        super().set_percussion_event(beat_key, position)
        self.flag('beat_change', beat_key)

    def set_percussion_instrument(self, line_offset: int, instrument: int) -> None:
        super().set_percussion_instrument(line_offset, instrument)
        for i in range(self.opus_beat_count):
            self.flag('beat_change', (9, line_offset, i))

    def unset(self, beat_key: BeatKey, position: List[int]) -> None:
        super().unset(beat_key, position)
        self.flag('beat_change', beat_key)

    def insert_beat(self, index: Optional[int] = None) -> None:
        if index is None:
            rindex = self.opus_beat_count - 1
        else:
            rindex = index

        super().insert_beat(index)

        self.flag('beat', (rindex, 'new'))

    def new_line(self,
            channel: int = 0,
            index: Optional[int] = None) -> None:
        super().new_line(channel, index)

        if index is None:
            line_index = len(self.channel_trees[channel]) - 1
        else:
            line_index = index

        self.flag('line', (channel, line_index, 'new'))

    def remove(self, beat_key: BeatKey, position: List[int]) -> None:
        super().remove(beat_key, position)
        self.flag('beat_change', beat_key)

    def remove_beat(self, index: None) -> None:
        if index is None:
            rindex = self.opus_beat_count - 1
        else:
            rindex = index

        super().remove_beat(index)

        self.flag('beat', (rindex, 'pop'))

    def remove_line(
            self,
            channel: int,
            index: Optional[int] = None) -> None:
        super().remove_line(channel, index)

        if index is None:
            index = len(self.channel_trees[channel])

        # Flag changes to cache
        self.flag('line', (channel, index, 'pop'))

File: kotlin/app/src/main/kotlin/radixulous/app/layer_flag.kt
package radixulous.app.opusmanager
import radixulous.app.structure.OpusTree
import radixulous.app.opusmanager.BeatKey
import radixulous.app.opusmanager.OpusEvent
import radixulous.app.opusmanager.OpusManagerBase

class UpdatesCache {
    var beat_pop: MutableList<Int> = mutableListOf()
    var beat_new: MutableList<Int> = mutableListOf()
    var beat_change: MutableList<BeatKey> = mutableListOf()
    var line_new: MutableList<Pair<Int, Int>> = mutableListOf()
    var line_pop: MutableList<Pair<Int, Int>> = mutableListOf()
    var line_init: MutableList<Pair<Int, Int>> = mutableListOf()

    public fun flag_beat_pop(index: Int) {
        this.beat_pop.add(index)
    }

    public fun flag_beat_new(index: Int) {
        this.beat_new.add(index)
    }

    public fun flag_beat_change(beat_key: BeatKey) {
        this.beat_change.add(beat_key)
    }

    public fun flag_line_pop(channel: Int, line_offset: Int) {
        this.line_pop.add(Pair(channel, line_offset))
    }

    public fun flag_line_new(channel: Int, line_offset: Int) {
        this.line_new.add(Pair(channel, line_offset))
    }

    public fun flag_line_init(channel: Int, line_offset: Int) {
        this.line_init.add(Pair(channel, line_offset))
    }
}

open class FlagLayer() : OpusManagerBase() {
    var cache = UpdatesCache()

    override fun replace_tree(beat_key: BeatKey, position: List<Int>, tree: OpusTree<OpusEvent>) {
        super.replace_tree(beat_key, position, tree)
        this.cache.flag_beat_change(beat_key)
    }

    override fun overwrite_beat(old_beat: BeatKey, new_beat: BeatKey) {
        super.overwrite_beat(old_beat, new_beat)
        this.cache.flag_beat_change(old_beat)
    }

    override fun insert_after(beat_key: BeatKey, position: List<Int>) {
        super.insert_after(beat_key, position)
        this.cache.flag_beat_change(beat_key)
    }

    override fun split_tree(beat_key: BeatKey, position: List<Int>, splits: Int) {
        super.split_tree(beat_key, position, splits)
        this.cache.flag_beat_change(beat_key)
    }

    override fun swap_channels(channel_a: Int, channel_b: Int) {
        super.swap_channels(channel_a, channel_b)
        var len_a = this.channel_trees[channel_a].size
        var len_b = this.channel_trees[channel_b].size
        for (i in 0 .. len_b - 1) {
            this.cache.flag_line_pop(channel_a, len_b - 1 - i)
        }

        for (i in 0 .. len_a - 1) {
            this.cache.flag_line_pop(channel_b, len_a - 1 - i)
        }

        for (i in 0 .. len_b - 1) {
            this.cache.flag_line_new(channel_b, i)
        }

        for (i in 0 .. len_a - 1) {
            this.cache.flag_line_new(channel_a, i)
        }
    }

    override fun new() {
        super.new()
        for (i in 0 .. this.opus_beat_count - 1) {
            this.cache.flag_beat_new(i)
        }
        for (i in 0 .. this.channel_trees.size - 1) {
            var channel = this.channel_trees[i]
            for (j in 0 .. channel.size - 1) {
                this.cache.flag_line_init(i, j)
            }
        }
    }
    override fun load(path: String) {
        super.load(path)
        for (i in 0 .. this.opus_beat_count - 1) {
            this.cache.flag_beat_new(i)
        }
        for (i in 0 .. this.channel_trees.size - 1) {
            var channel = this.channel_trees[i]
            for (j in 0 .. channel.size - 1) {
                this.cache.flag_line_init(i, j)
            }
        }
    }

    override fun set_event(beat_key: BeatKey, position: List<Int>, event: OpusEvent) {
        super.set_event(beat_key, position, event)
        this.cache.flag_beat_change(beat_key)
    }

    override fun set_percussion_event(beat_key: BeatKey, position: List<Int>) {
        super.set_percussion_event(beat_key, position)
        this.cache.flag_beat_change(beat_key)
    }

    override fun unset(beat_key: BeatKey, position: List<Int>) {
        super.unset(beat_key, position)
        this.cache.flag_beat_change(beat_key)
    }

    override fun insert_beat(index: Int?) {
        var rindex: Int
        if (index == null) {
            rindex = this.opus_beat_count - 1
        } else {
            rindex = index
        }

        super.insert_beat(index)

        this.cache.flag_beat_new(rindex)
    }

    override fun new_line(channel: Int, index: Int?) {
        super.new_line(channel, index)

        var line_index = if (index == null) {
            this.channel_trees[channel].size - 1
        } else {
            index
        }
        this.cache.flag_line_new(channel, line_index)
    }

    override fun remove(beat_key: BeatKey, position: List<Int>) {
        super.remove(beat_key, position)
        this.cache.flag_beat_change(beat_key)
    }

    override fun remove_beat(index: Int?) {
        var rindex = if (index == null) {
            this.opus_beat_count - 1
        } else {
            index
        }
        super.remove_beat(index)
        this.cache.flag_beat_pop(rindex)
    }

    override fun remove_line(channel: Int, index: Int?) {
        super.remove_line(channel, index)

        if (index == null) {
            this.cache.flag_line_pop(channel, this.channel_trees[channel].size)
        } else {
            this.cache.flag_line_pop(channel, index)
        }
    }
}
